File: leetcode/131-palindrome-partitioning.py
# https://leetcode.com/problems/palindrome-partitioning/ (top-interview-questions)

from typing import List

class Solution:
    def _is_palindrom(self, s):
        left, right = 0, len(s)-1
        while left < right:
            if s[left] != s[right]:
                return False
            left += 1
            right -= 1
        return True



    def partition(self, s: str) -> List[List[str]]:
        partitions = list()
        partition = list()
        def dfs(index):
            if index == len(s):
                partitions.append(partition[:])
                return
            
            for i in range(index, len(s)):
                if self._is_palindrom(s[index:i+1]):
                    partition.append(s[index:i+1])
                    dfs(i+1)
                    partition.pop()
        dfs(0)
        return partitions

File: leetcode/131-palindrome-partitioning.js
// https://leetcode.com/problems/palindrome-partitioning/ (top-interview-questions)

/**
 * @param {string} s
 * @return {string[][]}
 */
var partition = function(s) {


  const partitions = [];
  const partition = [];


  const isPalindrom = (str) => {

    let l = 0;
    let r = str.length - 1;

    while (l < r) {
      if (str[l] !== str[r]) {
        return false;
      }
      l += 1;
      r -= 1;
    }
    return true;
  };


  const generatePartitions = (i) => {
    if (i === s.length) {
      partitions.push([...partition]);
      return;
    }

    for (let j = i+1; j <= s.length; j++) {
      const subStr = s.slice(i, j);
      if (!isPalindrom(subStr)) {
        continue;
      }
      partition.push(subStr);
      generatePartitions(j);
      partition.pop();
    }

  }

  generatePartitions(0);

  return partitions;

};
